crates/core/tedge_mapper/src/mapper.rs

```diff
@@ -85,14 +89,121 @@ impl Mapper {
             let _ = self.output.send(init_message).await;
         }
 
-        while let Some(message) = &mut self.input.next().await {
-            let converted_messages = self.converter.convert(&message);
-            for converted_message in converted_messages.into_iter() {
-                let _ = self.output.send(converted_message).await;
-            }
+        // Detect and reconcile any alarms that this mapper may have missed while not running
+        self.reconcile_missed_alarms().await;
+
+        // Main loop processing messages
+        while let Some(message) = self.input.next().await {
+            self.process_message(message).await;
         }
+
         Ok(())
     }
+
+    /// Detect and reconcile any alarms that were raised/cleared while this mapper process was not running.
+    /// Reconciliation is done by maintaining an internal journal of all the alarms processed by this mapper,
+    /// which is compared against all the live alarms seen by the mapper on every startup.
+    ///
+    /// All the live alarms are received from tedge/alarms topic on startup.
+    /// Similarly, all the previously processed alarms are received from tedge-internal/alarms topic.
+    /// Reconiciliation detects the difference between these two sets, which are the missed messages.
+    ///
+    /// An alarm that is present in tedge-internal/alarms, but not in tedge/alarms topic
+    /// is assumed to have been cleared while the mapper process was down.
+    /// Similarly, an alarm that is present in tedge/alarms, but not in tedge-internal/alarms topic
+    /// is one that was raised while the mapper process was down.
+    /// An alarm present in both, if their payload is the same, is one that was already processed before the restart
+    /// and hence can be ignored during reconcicilation.
+    async fn reconcile_missed_alarms(&mut self) {
+        let mut pending_alarms_map = HashMap::new();
+        let mut old_alarms_map = HashMap::new();
+
+        let reconciliation_window = tokio::time::sleep(Duration::from_secs(3));
+        tokio::pin!(reconciliation_window); // Explicit pinning is required to use this sleep Future in a select! loop
+
+        loop {
+            tokio::select! {
+                // When the reconciliation window times out, break out of the select loop to stop caching alarm messages
+                _ = &mut reconciliation_window => {
+                    break;
+                }
+
+                // During the reconciliation window, cache all the alarm messages received from tedge/alarms and tedge-internal alarms topics.
+                // This window gives the mapper ample time to receive all the retained messages from both the topics.
+                Some(message) = self.input.next() => {
+                    if message.topic.name.starts_with("tedge/alarms") {
```

Comment:
Beware, that we are consuming *all* types of messaging here, but processing *only* those related to alarms.

One needs:
* either to process all non-alarm messages with `self.converter.convert(&message)`,
* or to use a different connection and session name for this reconciliation step.

I have a preference for the second alternative, because doing so:
* The reconciliation loop focuses on alarms.
* The main process loop doesn't need to subscribe to the `tedge-internal/alarms/` alarm topic.

Reply:
My idea was to use two connections in sequence. A first one that only reads alarms and internal alarms and stop when all the retained messages has been received. A second one that reads alarms and measurements. I agree that there is window where we can have a race.

I'm not convinced that a single connection is fool-proof. Sure, this remove the race condition on cleared alarms; but this also make things complicated/fragile. For instance, the current version of the code forget to process the measurements during the reconciliation loop.

Having two parallel connections seems better to me. No race condition on cleared alarms and no risk to have regressions when we update the alarm or measurement processing. We don't have to stop listening for internal topics , we can simple ignore them - as done is a solution with a single connection.